File: frontend/src/api/speciesExplorer.ts
import { request } from "./client";

import type {
    SpeciesExplorerFilters,
} from "../types/speciesFilters";

import type {
    SpeciesExplorerResponse,
} from "../types/species";

export async function getSpeciesExplorer(
    filters: SpeciesExplorerFilters = {},
): Promise<SpeciesExplorerResponse> {
    const params = new URLSearchParams();

    if (filters.search) {
        params.set("search", filters.search);
    }

    if (filters.orderId) {
        params.set("order_id", filters.orderId);
    }

    if (filters.familyId) {
        params.set("family_id", filters.familyId);
    }

    if (filters.countryId) {
        params.set("country_id", filters.countryId);
    }

    if (
        filters.latitude !== undefined &&
        filters.longitude !== undefined
    ) {
        params.set("latitude", filters.latitude.toString());
        params.set("longitude", filters.longitude.toString());
    }

    if (filters.radiusKm !== undefined) {
        params.set("radius_km", filters.radiusKm.toString());
    }

    if (filters.hotspotId) {
        params.set("hotspot_id", filters.hotspotId);
    }

    params.set("page", (filters.page ?? 1).toString());
    params.set("page_size", (filters.pageSize ?? 25).toString());

    return request(`/species/explorer?${params.toString()}`);
}

File: backend/app/schemas/species/species_explorer.py
from uuid import UUID

from pydantic import BaseModel


class SpeciesExplorerSpecies(BaseModel):
    id: UUID

    common_name: str
    scientific_name: str

    image_url: str | None = None
    thumbnail_url: str | None = None

    family_common_name: str | None = None
    order_common_name: str | None = None


class SpeciesExplorerFilters(BaseModel):
    search: str | None = None

    order_id: UUID | None = None
    family_id: UUID | None = None
    country_id: UUID | None = None

    latitude: float | None = None
    longitude: float | None = None
    radius_km: float | None = None

    hotspot_id: UUID | None = None

    page: int = 1
    page_size: int = 25


class SpeciesExplorerResponse(BaseModel):
    items: list[SpeciesExplorerSpecies]

    total: int
    page: int
    page_size: int